File: interfaces/python/tests/unittests/update_matrices_tests.py
# Test osqp python module
import osqp
import numpy as np
import scipy as sp
from scipy import sparse

# Unit Test
import unittest
import numpy.testing as nptest


class update_matrices_tests(unittest.TestCase):

    def setUp(self):
        # Simple QP problem
        sp.random.seed(1)

        self.n = 5
        self.m = 8
        p = 0.7

        Pt = sparse.random(self.n, self.n, density=p)
        Pt_new = Pt.copy()
        Pt_new.data += 0.1 * np.random.randn(Pt.nnz)

        self.P = Pt.T.dot(Pt).tocsc() + sparse.eye(self.n).tocsc()
        self.P_new = Pt_new.T.dot(Pt_new).tocsc() + sparse.eye(self.n).tocsc()
        self.q = np.random.randn(self.n)
        self.A = sparse.random(self.m, self.n, density=p).tocsc()
        self.A_new = self.A.copy()
        self.A_new.data += np.random.randn(self.A_new.nnz)
        self.l = np.zeros(self.m)
        self.u = 30 + np.random.randn(self.m)
        self.opts = {'eps_abs': 1e-08,
                     'eps_rel': 1e-08,
                     'verbose': False}
        self.model = osqp.OSQP()
        self.model.setup(P=self.P, q=self.q, A=self.A, l=self.l, u=self.u,
                         **self.opts)

    def test_solve(self):
        # Solve problem
        res = self.model.solve()

        # Assert close
        nptest.assert_array_almost_equal(res.x,
            np.array([0.85459329, 0.73472366, 0.06156, -0.06095794, -0.96167612]))
        nptest.assert_array_almost_equal(res.y,
            np.array([0., 0., 0., -2.32275314, 0., -0.93213354, 0., -0.8939565]))
        nptest.assert_array_almost_equal(res.info.obj_val, -1.5116431929127323)

    def test_update_P(self):
        # Update matrix P
        Pnew_triu = sparse.triu(self.P_new).tocsc()
        Px = Pnew_triu.data
        Px_idx = np.arange(Pnew_triu.nnz)
        self.model.update(Px=Px, Px_idx=Px_idx)
        res = self.model.solve()

        # Assert close
        nptest.assert_array_almost_equal(res.x,
            np.array([0.79105808, 0.68008954, -0.00974931, -0.00569589, -0.92142316]))
        nptest.assert_array_almost_equal(res.y,
            np.array([0., -0.0977256, 0., -2.3051196, 0., -0.84705904, 0., -0.9014214]))
        nptest.assert_array_almost_equal(res.info.obj_val, -1.40901946656505)

    def test_update_P_allind(self):
        # Update matrix P
        Pnew_triu = sparse.triu(self.P_new).tocsc()
        Px = Pnew_triu.data
        self.model.update(Px=Px)
        res = self.model.solve()

        # Assert close
        nptest.assert_array_almost_equal(res.x,
            np.array([0.79105808, 0.68008954, -0.00974931, -0.00569589, -0.92142316]))
        nptest.assert_array_almost_equal(res.y,
            np.array([0., -0.0977256, 0., -2.3051196, 0., -0.84705904, 0., -0.9014214]))
        nptest.assert_array_almost_equal(res.info.obj_val, -1.40901946656505)

    def test_update_A(self):
        # Update matrix A
        Ax = self.A_new.data
        Ax_idx = np.arange(self.A_new.nnz)
        self.model.update(Ax=Ax, Ax_idx=Ax_idx)
        res = self.model.solve()

        # Assert close
        nptest.assert_array_almost_equal(res.x,
            np.array([0.44557958, 0.11209195, 0.22051994, -0.78051077, -0.01697192]))
        nptest.assert_array_almost_equal(res.y,
            np.array([-1.97318457, -1.43719371, 0., -0.05364337, -1.3354648, 0., 0., 0.]))
        nptest.assert_array_almost_equal(res.info.obj_val, -0.79990427087463)

    def test_update_A_allind(self):
        # Update matrix A
        Ax = self.A_new.data
        self.model.update(Ax=Ax)
        res = self.model.solve()

        # Assert close
        nptest.assert_array_almost_equal(res.x,
            np.array([0.44557958, 0.11209195, 0.22051994, -0.78051077, -0.01697192]))
        nptest.assert_array_almost_equal(res.y,
            np.array([-1.97318457, -1.43719371, 0., -0.05364337, -1.3354648, 0., 0., 0.]))
        nptest.assert_array_almost_equal(res.info.obj_val, -0.79990427087463)

    def test_update_P_A_indP_indA(self):
        # Update matrices P and A
        Pnew_triu = sparse.triu(self.P_new).tocsc()
        Px = Pnew_triu.data
        Px_idx = np.arange(Pnew_triu.nnz)
        Ax = self.A_new.data
        Ax_idx = np.arange(self.A_new.nnz)
        self.model.update(Px=Px, Px_idx=Px_idx, Ax=Ax, Ax_idx=Ax_idx)
        res = self.model.solve()

        # Assert close
        nptest.assert_array_almost_equal(res.x,
            np.array([0.45599336, 0.11471169, 0.22567378, -0.80654725, -0.01778191]))
        nptest.assert_array_almost_equal(res.y,
            np.array([-1.76495387, -1.44638239, 0., 0., -1.28476339, 0., 0., 0.]))
        nptest.assert_array_almost_equal(res.info.obj_val, -0.8249598023368026)

    def test_update_P_A_indP(self):
        # Update matrices P and A
        Pnew_triu = sparse.triu(self.P_new).tocsc()
        Px = Pnew_triu.data
        Px_idx = np.arange(Pnew_triu.nnz)
        Ax = self.A_new.data
        self.model.update(Px=Px, Px_idx=Px_idx, Ax=Ax)
        res = self.model.solve()

        # Assert close
        nptest.assert_array_almost_equal(res.x,
            np.array([0.45599336, 0.11471169, 0.22567378, -0.80654725, -0.01778191]))
        nptest.assert_array_almost_equal(res.y,
            np.array([-1.76495387, -1.44638239, 0., 0., -1.28476339, 0., 0., 0.]))
        nptest.assert_array_almost_equal(res.info.obj_val, -0.8249598023368026)

    def test_update_P_A_indA(self):
        # Update matrices P and A
        Pnew_triu = sparse.triu(self.P_new).tocsc()
        Px = Pnew_triu.data
        Ax = self.A_new.data
        Ax_idx = np.arange(self.A_new.nnz)
        self.model.update(Px=Px, Ax=Ax, Ax_idx=Ax_idx)
        res = self.model.solve()

        # Assert close
        nptest.assert_array_almost_equal(res.x,
            np.array([0.45599336, 0.11471169, 0.22567378, -0.80654725, -0.01778191]))
        nptest.assert_array_almost_equal(res.y,
            np.array([-1.76495387, -1.44638239, 0., 0., -1.28476339, 0., 0., 0.]))
        nptest.assert_array_almost_equal(res.info.obj_val, -0.8249598023368026)

    def test_update_P_A_allind(self):
        # Update matrices P and A
        Pnew_triu = sparse.triu(self.P_new).tocsc()
        Px = Pnew_triu.data
        Ax = self.A_new.data
        self.model.update(Px=Px, Ax=Ax)
        res = self.model.solve()

        # Assert close
        nptest.assert_array_almost_equal(res.x,
            np.array([0.45599336, 0.11471169, 0.22567378, -0.80654725, -0.01778191]))
        nptest.assert_array_almost_equal(res.y,
            np.array([-1.76495387, -1.44638239, 0., 0., -1.28476339, 0., 0., 0.]))
        nptest.assert_array_almost_equal(res.info.obj_val, -0.8249598023368026)

File: interfaces/matlab/unittests/update_matrices_tests.m
classdef update_matrices_tests < matlab.unittest.TestCase
    %TEST_BASIC_QP Solve Basic QP Problem

    properties
        P
        P_new
        q
        A
        A_new
        u
        l
        m
        n
        solver
        options
        tol
    end

    methods(TestMethodSetup)
        function setup_problem(testCase)
            rng(1)

            testCase.n = 5;
            testCase.m = 8;
            p = 0.7;

            Pt = sprandn(testCase.n, testCase.n, p);
            Pt_new = Pt;
            Pt_new(find(Pt)) = Pt(find(Pt)) + 0.1*randn(nnz(Pt), 1);

            % Create Problem
            testCase.P = Pt'*Pt + speye(testCase.n);
            testCase.P_new = Pt_new'*Pt_new + speye(testCase.n);
            testCase.q = randn(testCase.n, 1);
            testCase.A = sprandn(testCase.m, testCase.n, p);
            testCase.A_new = testCase.A;
            testCase.A_new(find(testCase.A)) = testCase.A(find(testCase.A)) + randn(nnz(testCase.A), 1);
            testCase.l = zeros(testCase.m, 1);
            testCase.u = 30 + randn(testCase.m, 1);

            % Setup solver
            testCase.solver = osqp;
            testCase.solver.setup(testCase.P, testCase.q, testCase.A, ...
                testCase.l, testCase.u, 'verbose', false, 'eps_rel', 1e-7, 'eps_abs', 1e-07, 'polish', true);

            % Setup tolerance
            testCase.tol = 1e-04;

        end
    end

    methods (Test)
        function test_solve(testCase)
            % Solve with OSQP
            results = testCase.solver.solve();

            % Check if they are close
            testCase.verifyEqual(results.x, ...
                [-0.0000; -0.0281; 0.2292; 0.0000; -0.0000], 'AbsTol',testCase.tol)
% Dual solution returned might be different
%             testCase.verifyEqual(results.y, ...
%                 [-1.5643; 0; -1.3562; -0.9056; -0.4113; 0; 0; 0], 'AbsTol',testCase.tol)
            testCase.verifyEqual(results.info.obj_val, -0.0400, 'AbsTol', testCase.tol)

        end

        function test_update_P(testCase)
            % Update matrix P
            Pnew_triu = triu(testCase.P_new);
            Px = nonzeros(Pnew_triu);
            Px_idx = (1:nnz(Pnew_triu))';
            testCase.solver.update('Px', Px, 'Px_idx', Px_idx);

            % Solve again
            results = testCase.solver.solve();

            % Check if they are close
            testCase.verifyEqual(results.x, ...
                [-0.0000; -0.0261; 0.2129; 0.0000; -0.0000], 'AbsTol',testCase.tol)
% Dual solution returned might be different
%             testCase.verifyEqual(results.y, ...
%                 [-1.5616; 0; -1.3631; -0.8931; -0.4198; 0; 0; 0], 'AbsTol',testCase.tol)
            testCase.verifyEqual(results.info.obj_val, -0.0372, 'AbsTol', testCase.tol)

        end
        
        function test_update_P_allind(testCase)
            % Update matrix P
            Pnew_triu = triu(testCase.P_new);
            Px = nonzeros(Pnew_triu);
            testCase.solver.update('Px', Px);

            % Solve again
            results = testCase.solver.solve();

            % Check if they are close
            testCase.verifyEqual(results.x, ...
                [-0.0000; -0.0261; 0.2129; 0.0000; -0.0000], 'AbsTol',testCase.tol)
% Dual solution returned might be different
%             testCase.verifyEqual(results.y, ...
%                 [-1.5616; 0; -1.3631; -0.8931; -0.4198; 0; 0; 0], 'AbsTol',testCase.tol)
            testCase.verifyEqual(results.info.obj_val, -0.0372, 'AbsTol', testCase.tol)

        end

        function test_update_A(testCase)
            % Update matrix A
            Ax = nonzeros(testCase.A_new);
            Ax_idx = (1:nnz(testCase.A_new))';
            testCase.solver.update('Ax', Ax, 'Ax_idx', Ax_idx);

            % Solve again
            results = testCase.solver.solve();

            % Check if they are close
            testCase.verifyEqual(results.x, ...
                [-0.0398; -0.0761; -0.0292; -0.0000; -0.0199], 'AbsTol',testCase.tol)
% Dual solution returned might be different
%             testCase.verifyEqual(results.y, ...
%                 [-1.2329; 0; 0; 0; -0.3052; -0.4643; -0.0151; 0], 'AbsTol',testCase.tol)
            testCase.verifyEqual(results.info.obj_val, -0.0093, 'AbsTol', testCase.tol)

        end
        
        function test_update_A_allind(testCase)
            % Update matrix A
            Ax = nonzeros(testCase.A_new);
            testCase.solver.update('Ax', Ax);

            % Solve again
            results = testCase.solver.solve();

            % Check if they are close
            testCase.verifyEqual(results.x, ...
                [-0.0398; -0.0761; -0.0292; -0.0000; -0.0199], 'AbsTol',testCase.tol)
% Dual solution returned might be different
%             testCase.verifyEqual(results.y, ...
%                 [-1.2329; 0; 0; 0; -0.3052; -0.4643; -0.0151; 0], 'AbsTol',testCase.tol)
            testCase.verifyEqual(results.info.obj_val, -0.0093, 'AbsTol', testCase.tol)

        end
        
        function test_update_P_A_indP_indA(testCase)
            % Update matrices P and A
            Pnew_triu = triu(testCase.P_new);
            Px = nonzeros(Pnew_triu);
            Px_idx = (1:nnz(Pnew_triu))';
            Ax = nonzeros(testCase.A_new);
            Ax_idx = (1:nnz(testCase.A_new))';
            testCase.solver.update('Px', Px, 'Px_idx', Px_idx, 'Ax', Ax, 'Ax_idx', Ax_idx);

            % Solve again
            results = testCase.solver.solve();

            % Check if they are close
            testCase.verifyEqual(results.x, ...
                [-0.0439; -0.0840; -0.0322; 0.0000; -0.0219], 'AbsTol',testCase.tol)
% Dual solution returned might be different
%             testCase.verifyEqual(results.y, ...
%                 [-1.2386; 0; 0; 0; -0.3105; -0.4430; -0.0112; 0], 'AbsTol',testCase.tol)
            testCase.verifyEqual(results.info.obj_val, -0.0103, 'AbsTol', testCase.tol)

        end
        
        function test_update_P_A_indP(testCase)
            % Update matrices P and A
            Pnew_triu = triu(testCase.P_new);
            Px = nonzeros(Pnew_triu);
            Px_idx = (1:nnz(Pnew_triu))';
            Ax = nonzeros(testCase.A_new);
            testCase.solver.update('Px', Px, 'Px_idx', Px_idx, 'Ax', Ax);

            % Solve again
            results = testCase.solver.solve();

            % Check if they are close
            testCase.verifyEqual(results.x, ...
                [-0.0439; -0.0840; -0.0322; 0.0000; -0.0219], 'AbsTol',testCase.tol)
% Dual solution returned might be different
%             testCase.verifyEqual(results.y, ...
%                 [-1.2386; 0; 0; 0; -0.3105; -0.4430; -0.0112; 0], 'AbsTol',testCase.tol)
            testCase.verifyEqual(results.info.obj_val, -0.0103, 'AbsTol', testCase.tol)

        end
        
        function test_update_P_A_indA(testCase)
            % Update matrices P and A
            Pnew_triu = triu(testCase.P_new);
            Px = nonzeros(Pnew_triu);
            Ax = nonzeros(testCase.A_new);
            Ax_idx = (1:nnz(testCase.A_new))';
            testCase.solver.update('Px', Px, 'Ax', Ax, 'Ax_idx', Ax_idx);

            % Solve again
            results = testCase.solver.solve();

            % Check if they are close
            testCase.verifyEqual(results.x, ...
                [-0.0439; -0.0840; -0.0322; 0.0000; -0.0219], 'AbsTol',testCase.tol)
% Dual solution returned might be different
%             testCase.verifyEqual(results.y, ...
%                 [-1.2386; 0; 0; 0; -0.3105; -0.4430; -0.0112; 0], 'AbsTol',testCase.tol)
            testCase.verifyEqual(results.info.obj_val, -0.0103, 'AbsTol', testCase.tol)

        end
        
        function test_update_P_A_allind(testCase)
            % Update matrices P and A
            Pnew_triu = triu(testCase.P_new);
            Px = nonzeros(Pnew_triu);
            Ax = nonzeros(testCase.A_new);
            testCase.solver.update('Px', Px, 'Ax', Ax);

            % Solve again
            results = testCase.solver.solve();

            % Check if they are close
            testCase.verifyEqual(results.x, ...
                [-0.0439; -0.0840; -0.0322; 0.0000; -0.0219], 'AbsTol',testCase.tol)
% Dual solution returned might be different
%             testCase.verifyEqual(results.y, ...
%                 [-1.2386; 0; 0; 0; -0.3105; -0.4430; -0.0112; 0], 'AbsTol',testCase.tol)
            testCase.verifyEqual(results.info.obj_val, -0.0103, 'AbsTol', testCase.tol)

        end
    
    end

end
